mlkem/asm/aarch64/rej_uniform_asm_clean.S

```diff
@@ -0,0 +1,585 @@
+    iter                        .req x10
+
+    rec_idx_0                   .req w12
+    rec_idx_1                   .req w13
+    rec_idx_2                   .req w14
+    rec_idx_3                   .req w15
+
+    ctr0                        .req w12
+    ctr1                        .req w13
+    ctr2                        .req w14
+    ctr3                        .req w15
+
+    /* Vector registers */
+
+    buf0                        .req v0
+    buf1                        .req v1
+    buf2                        .req v2
+
+    tmp0                        .req v4
+    tmp1                        .req v5
+    tmp2                        .req v6
+    tmp3                        .req v7
+
+    sign0                       .req v4
+    sign1                       .req v5
+    sign2                       .req v6
+    sign3                       .req v7
+
+    val0                        .req v16
+    val0q                       .req q16
+    val1                        .req v17
+    val1q                       .req q17
+    val2                        .req v18
+    val2q                       .req q18
+    val3                        .req v19
+    val3q                       .req q19
+
+    t0                          .req s20
+    t1                          .req s21
+    t2                          .req s22
+    t3                          .req s23
+
+    table0                      .req v24
+    table0q                     .req q24
+    table1                      .req v25
+    table1q                     .req q25
+    table2                      .req v26
+    table2q                     .req q26
+    table3                      .req v27
+    table3q                     .req q27
+
+    kyber_q                     .req v30
+    bits                        .req v31
+    bits_q                      .req q31
+
+
+    /*
+    * Preserve registers and restore them at the end of the function.
+    * Stack alignment must be 16 bytes.
+    * Stack size: 0x270 = 16 * 3 (preserve registers + place_holder) + sizeof(int16_t) * (KYBER_N + 32)
+    * Memory layout from the bottom of the stack to the top:
+    * 0x00:   x8            |  place_holder
+    * 0x10:   x16           |  x17
+    * 0x20:   x29           |  x30
+    * Start at 0x030 to the end at 0x220:   KYBER_N * sizeof(int16_t)
+    * 0x230 -> 0x270: 64 = sizeof(int16_t) * 32 bytes padding to avoid overflow in the `str` sequence.
+    * The 4 `str` sequences at the end of `loop48` can write from 0 bytes (4*0) to 64 bytes (4*16).
+    * However, we only copy the maximum amount of `len` stored in x4 from stack to `x0`.
+    */
+    sub     sp, sp, #0x270
+    str     x8, [sp]
+    stp     x16, x17, [sp, #16]
+    stp     x29, x30, [sp, #32]
+
+    ASM_LOAD(bit_table, bit_table_data)
+    ASM_LOAD(table_idx, table_data)
+
+    /* Vectorize code start */
+
+    ldr         bits_q, [bit_table]
+    movz        tmp, #3329
+    dup         kyber_q.8h, tmp
+    round_48x   bound, buflen
```

Comment:
Can it ever happen that bound is zero here,  i.e., buflen < 48?

Reply:
yes it can, the ASM version is batch sampling, which consumes 48 (loop48) or 24 (loop24) bytes at a time. When the `buflen` input is larger than 48, it goes to loop48, if it less than 48, it goes to loop24, if it less than 24, it goes to the scalar version.

Reply:
But the 48-loop is never skipped?

Reply:
I forget to add the check, it should check first then execute. I'm adding it now.

Reply:
@cothan I'm going over everything right now and fixing it, no need to take action from your end.